demos/udp_server_demo.cc

```diff
@@ -45,38 +63,73 @@ class udp_server {
+                        _n_sent++;
+                    }).get();
+                } catch (seastar::gate_closed_exception&) {
+                    std::cout << "Gate closed" << std::endl;
+                    break;
+                } catch (seastar::broken_promise& e) {
+                    // Note - in case the other end closed the session this can happen.
+                    std::cerr << e.what() << std::endl;
+                }
+            }
+
+            if (!_chan.is_closed())
+                _chan.close();
+            _stats_timer.cancel();
         });
+
+        _stopper = std::make_unique<seastar::future<>>(std::move(stopper));
     }
-    // FIXME: we should properly tear down the service here.
-    future<> stop() {
-        return make_ready_future<>();
+
+    seastar::future<> stop() {
+        // Note that we need to call `_chan.close` here in case we are not getting any packets.
+        // In such a case the receiver will just hang and won't be terminated.
+        // By actively closing the channel here we will trigger `broken_promise` on the receiver.
+        _chan.close();
+
+        // Note that it is not enough to just close the gate here.
+        // If we just wait on `_gate.close` and ignore the stopper we will have a race on our hands.
+        // If the receiver thread is not currently within the gate (i.e gate.close will return) it doesn't mean
+        // it was terminated. If we return from the stop call before that we reach a scenario in which the 
+        // object is destroyed but the thread wasn't terminated yet -> use_after_free issue.
+        return seastar::when_all(std::move(_gate.close()), std::move(*_stopper.get())).discard_result();
     }
 };
 
 namespace bpo = boost::program_options;
 
 int main(int ac, char ** av) {
-    app_template app;
+    seastar::app_template app;
+    
     app.add_options()
         ("port", bpo::value<uint16_t>()->default_value(10000), "UDP server port") ;
-    return app.run_deprecated(ac, av, [&] {
+    return app.run(ac, av, [&]() -> seastar::future<> {
+        seastar_apps_lib::stop_signal stop_signal;
         auto&& config = app.configuration();
         uint16_t port = config["port"].as<uint16_t>();
-        auto server = new distributed<udp_server>;
+        auto server = std::make_unique<seastar::distributed<udp_server>>();
+
         // Run server in background.
-        (void)server->start().then([server = std::move(server), port] () mutable {
-            engine().at_exit([server] {
-                return server->stop();
-            });
-            return server->invoke_on_all(&udp_server::start, port);
-        }).then([port] {
-            std::cout << "Seastar UDP server listening on port " << port << " ...\n";
-        });
+        co_await server->start();
+
+        std::cout << "Start listening on all shards...\n";
+        co_await server->invoke_on_all(&udp_server::start, port);
+
```

Comment:
This won't work on C++17, that doesn't have coroutines. It's customary to use threads (see seastar::async for an easy way to create one) in initialization code.